### PR TITLE
Fix: frame dependency calculation

### LINKED ISSUE #75: Bug in Timing/Frame Dependencies
Hot on the heels of #73 , another bug has been found in how frame blocking is calculated. Two examples are given here, with further explanation below. Can you spot the bug?

# Example 1

Hint: there is an additional frame defined in the program, `0 1 "cz"`, which is not used by any of the instructions.

graphviz (4)

# Example 2

graphviz (5)

The problem here is that the change in #73 properly fixed how dependencies are calculated _after_ a blocking instruction, but that blocking instruction _must itself block on previous instructions which overlap in frames_, and that is not happening.

Specifically/formally:

1. An instruction which _uses_ a frame must depend on the most-recent _user_ **and** _blockers_ of the frame, if any. Doing so clears all blockers of the frame and this instruction becomes the most-recent user.
2. An instruction which _blocks_ but does not _use_ a frame must depend on the most-recent user of the frame. Doing so leaves the most-recent user in place and adds this instruction to a list of most-recent blockers.

Blocking prevents simultaneous use but allows simultaneous blocking.

This approach is akin to a single-writer, multiple-reader lock or queue, where blocking:reading::using:writing.

### PR DESCRIPTION
**Pull request by kalzoo**
Another fix, back to back with #73.

Closes #75 

This takes the approach of tracking the instructions which _use_ and _block_ a frame separately, solving the problems described in #75. It's more complex, but should still remain readable and maintainable. This PR includes a couple more snapshot test cases to illustrate the differences and prevent regression.

### CODE REVIEW

**Comment by genos** [comment]
Issue-linking-wise, @kalzoo, is this aimed at #75?

File: src/program/graph.rs [hunk 1]
```diff
@@ -200,6 +200,75 @@ pub struct InstructionBlock {
     pub terminator: BlockTerminator,
 }
 
+/// PreviousNodes is a structure which helps maintain ordering among instructions which operate on a given frame.
```
Comment (genos) [inline src/program/graph.rs:203]: `PreviousNodes` _are_ a structure?

j/k

File: src/program/graph.rs [hunk 2]
```diff
@@ -233,24 +301,47 @@ impl InstructionBlock {
...
-                            .unwrap_or(&ScheduledGraphNode::BlockStart);
-                        add_dependency!(graph, *previous_node_id => node, ExecutionDependency::ReferenceFrame);
+                        let previous_node_ids = last_instruction_by_frame
+                            .entry(frame.clone())
+                            .or_insert(PreviousNodes {
+                                using: None,
+                                blocking: vec![ScheduledGraphNode::BlockStart]
+                                    .into_iter()
+                                    .collect(),
+                            })
```
Comment (genos) [inline src/program/graph.rs:324]:
```rust
.or_default()
```

File: src/program/graph.rs [hunk 3]
```diff
@@ -233,24 +301,47 @@ impl InstructionBlock {
...
                     for frame in blocked_frames {
-                        last_instruction_by_frame.insert(frame.clone(), node);
+                        if let Some(previous_node_id) = last_instruction_by_frame
+                            .entry(frame.clone())
+                            .or_insert(PreviousNodes {
+                                using: None,
+                                blocking: vec![ScheduledGraphNode::BlockStart]
+                                    .into_iter()
+                                    .collect(),
+                            })
```
Comment (genos) [inline src/program/graph.rs:340]:
```rust
.or_default()
```

File: src/program/graph.rs [hunk 4]
```diff
@@ -233,24 +301,47 @@ impl InstructionBlock {
                     Ok(())
                 }
                 InstructionRole::RFControl => {
-                    let used_frames = program
+                    let used_frames: HashSet<&FrameIdentifier> = program
```
Comment (genos) [inline src/program/graph.rs:304]: `get_frames_for_instruction` returns an `Option<Vec<<&FrameIdentifier>>`, but its only uses always convert that to a `HashSet<&FrameIdentifier>`. Would it make sense to DRY this out and change the function definition?

File: src/program/graph.rs [hunk 5]
```diff
@@ -295,8 +386,10 @@ impl InstructionBlock {
         // does not terminate until these are complete
         add_dependency!(graph, last_classical_instruction => ScheduledGraphNode::BlockEnd, ExecutionDependency::StableOrdering);
 
-        for (_, last_instruction) in last_instruction_by_frame {
-            add_dependency!(graph, last_instruction => ScheduledGraphNode::BlockEnd, ExecutionDependency::ReferenceFrame);
+        for (_, last_instructions) in last_instruction_by_frame {
```
Comment (genos) [inline src/program/graph.rs:389]:
```rust
for last_instruction in last_instruction_by_frame.values()
```

File: src/program/graph.rs [hunk 6]
```diff
@@ -200,6 +200,75 @@ pub struct InstructionBlock {
...
+        result
+    }
+
+    /// Register a node as blocking a frame, and return the instructions on which it should depend/wait for scheduling (if any).
+    ///
+    /// A node which blocks a frame will block on any previous user of the frame, but not concurrent blockers.
+    ///
+    /// If the frame is currently blocked by other nodes, it will add itself to the list of blockers,
+    /// much like a reader in a read-write lock.
+    pub fn register_blocker(&mut self, node: ScheduledGraphNode) -> Option<ScheduledGraphNode> {
```
Comment (genos) [inline src/program/graph.rs:258]: I can't think of a better name, so maybe ignore this, but I think a more suggestive function name that says something about returning `self.using` might be helpful.

I suppose I have a similar (-ly unhelpful) note for `register_user`'s naming, too.

**Review by genos — COMMENTED** [review]
Some requests for shorter code 🏌️, a suggestion about changing the return type of a function, but otherwise LGTM

File: src/program/graph.rs (hunk 2 above)
Comment (kalzoo) [reply src/program/graph.rs:324]: Ah right! I added the default impl then forgot to use it

File: src/program/graph.rs (hunk 4 above)
Comment (kalzoo) [reply src/program/graph.rs:304]: Yes - although that's a breaking change - _although_ that's fine these days.

**Comment by kalzoo** [comment]
Note: this fails the example cited in [this issue](https://github.com/quil-lang/quil/issues/59); if the spec remains unchanged, this will have to be modified.

File: src/program/graph.rs (hunk 6 above)
Comment (kalzoo) [reply src/program/graph.rs]: Changed - what do you think?

File: src/program/mod.rs [hunk 7]
```diff
@@ -130,15 +130,10 @@ impl Program {
...
+    ) -> Option<HashSet<&'a FrameIdentifier>> {
         instruction
             .get_frame_match_condition(include_blocked)
-            .map(|condition| {
-                self.frames
-                    .get_matching_keys(condition)
-                    .into_iter()
-                    .collect()
-            })
+            .map(|condition| self.frames.get_matching_keys(condition))
```
Comment (notmgsk) [inline src/program/mod.rs:136]: How come this doesn't need a `.collect()`? 🤔

File: src/program/graph.rs [hunk 8]
```diff
@@ -200,6 +200,81 @@ pub struct InstructionBlock {
...
+            blocking: vec![ScheduledGraphNode::BlockStart].into_iter().collect(),
+        }
+    }
+}
+
+impl PreviousNodes {
+    /// Register a node as using a frame, and return the instructions on which it should depend/wait for scheduling (if any).
+    ///
+    /// A node which uses a frame will block on any previous user or blocker of the frame, much like a writer in a read-write lock.
+    pub fn get_dependencies_for_next_user(
```
Comment (notmgsk) [inline src/program/graph.rs:243]: Do these need to be `pub`?

**Review by notmgsk — APPROVED** [review]
Nice. Took me a hot minute to understand the logic, but that might be because I've been awake since 5AM. Only question is: do the `PreviousNodes` methods need to be public?

File: src/program/mod.rs (hunk 7 above)
Comment (genos) [reply src/program/mod.rs:136]: Because `get_matching_keys` already returns a `HashSet`?

That said, you might be able to shorten this to `.map(self.frames.get_matching_keys)`.

File: src/program/graph.rs (hunk 6 above)
Comment (genos) [reply src/program/graph.rs]: 👍

File: src/program/graph.rs (hunk 8 above)
Comment (kalzoo) [reply src/program/graph.rs]: nope, good call (although the struct itself is not pub)

File: src/program/mod.rs (hunk 7 above)
Comment (kalzoo) [reply src/program/mod.rs:136]: Compiler doesn't like passing that function to `map`, although I'm not sure why.
